### PR TITLE
Added required new methods for ha setups with remote nodes

### PR DESCRIPTION
**Pull request by tboerger**
**Depends on https://github.com/crowbar/crowbar-core/pull/193**

### CODE REVIEW

File: crowbar_framework/app/models/pacemaker_service_object.rb [hunk 1]
```diff
@@ -172,7 +218,7 @@ def allocate_virtual_ips_for_any_cluster_in_networks(elements, networks)
     new_allocation = false
 
     elements.each do |element|
-      if PacemakerServiceObject.is_cluster? element
+      if PacemakerServiceObject.is_cluster?(element) || PacemakerServiceObject.is_remotes?(element)
```
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Actually I don't think we want this to work on remotes.

File: crowbar_framework/app/models/pacemaker_service_object.rb [hunk 2]
```diff
@@ -208,7 +254,7 @@ def prepare_role_for_cluster_vip_networks(role, elements, networks)
     dirty = false
 
     elements.each do |element|
-      next unless PacemakerServiceObject.is_cluster? element
+      next unless PacemakerServiceObject.is_cluster?(element) && PacemakerServiceObject.is_remotes?(element)
```
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Same here, my gut feeling is that we don't want this on remotes.

File: crowbar_framework/app/models/pacemaker_service_object.rb [hunk 3]
```diff
@@ -142,7 +188,7 @@ def role_expand_elements(role, role_name)
   #     taken). This information can be used to know when a DNS sync is
   #     required or not.
   def allocate_virtual_ips_for_cluster_in_networks(cluster, networks)
-    if networks.nil? || networks.empty? || !PacemakerServiceObject.is_cluster?(cluster)
+    if networks.nil? || networks.empty? || !PacemakerServiceObject.is_cluster?(cluster) || !PacemakerServiceObject.is_remotes?(cluster)
```
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Same here (in short: virtual IPs should not be for remotes).

**Comment by vuntz** [comment]
Marking as WIP because of the TODO (hard-coded clusters).

File: crowbar_framework/app/models/pacemaker_service_object.rb [hunk 4]
```diff
@@ -50,39 +43,92 @@ def available_clusters
...
-      "cluster-#{name.gsub("_", "-")}.#{ChefObject.cloud_domain}"
-    end
-
     def cluster_vhostname_from_element(element)
-      if is_cluster? element
+      case
+      when is_remotes?(element)
+        remotes_vhostname_from_name(cluster_name(element))
+      when is_cluster?(element)
         cluster_vhostname_from_name(cluster_name(element))
```
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Similarly to the VIP comments, I'm not sure we care about a vhostname for remotes.

**Comment by vuntz** [comment]
Did some changes in https://github.com/tboerger/crowbar-ha/pull/1.

**Comment by tboerger** [comment]
@vuntz I merged your pr, so the source is updated now.

**Comment by vuntz** [comment]
I forgot, but +1.

File: crowbar_framework/app/models/pacemaker_service_object.rb [hunk 5]
```diff
@@ -81,8 +77,53 @@ def cluster_vhostname_from_name(name)
...
+      end
+    end
+
+    # Returns: name of the barclamp and of the proposal for this cluster
+    def cluster_get_barclamp_and_proposal(element)
+      case
+      when is_remotes?(element)
+        ["pacemaker", cluster_name(element)]
+      when is_cluster?(element)
+        ["pacemaker", cluster_name(element)]
```
Comment (jsuchome) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Is the result really the same for `is_cluster?` and `is_remotes?`

Comment (jsuchome) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Ah, `cluster_name` is different in those cases. But why do we need 2 `when` branches here?

File: crowbar_framework/app/models/pacemaker_service_object.rb [hunk 6]
```diff
@@ -50,26 +43,29 @@ def available_clusters
...
     end
 
-    # Returns: name of the barclamp and of the proposal for this cluster
-    def cluster_get_barclamp_and_proposal(element)
-      if is_cluster?(element)
-        ["pacemaker", cluster_name(element)]
-      else
-        [nil, nil]
+    # Returns: List of available clusters including remotes
+    def available_clusters_including_remotes
```
Comment (jsuchome) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: The name is somewhat confusing. I thought it's gonna return all available clusters, so those which has remotes as well as those who have them.

Isn't it rather `available_remotes`, `all_remotes_from_clusters`  or something like that?

Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: Agreed, the naming is a bit confusing. I would be fine with `available_remotes` (needs a change in the crowbar-core PR too).

File: crowbar_framework/app/models/pacemaker_service_object.rb (hunk 5 above)
Comment (vuntz) [inline crowbar_framework/app/models/pacemaker_service_object.rb]: I guess for consistency with some other method. I'm fine if this gets changed.

**Comment by jsuchome** [comment]
+1

**Comment by tboerger** [comment]
So merge this?

**Comment by vuntz** [comment]
> So merge this?

Waiting for the gating job, I guess.

**Comment by vuntz** [comment]
Gating fails due to https://trello.com/c/30w2SCA0/128-c6-ha-waiting-for-testvm-booted-and-ping-returned-timed-out-in-testsetup-step (tempest passed, though)

Given that the code change is on the crowbar side of things, and actually unused yet (is used by other PR), I think we're good enough for merging.

**Pull request merged**